RFC-0004-expression-eval.md

```diff
@@ -0,0 +1,80 @@
+
+## Background
+
+Unfenced functions are functions which execute within the same process as the underlying evaluation engine.  Typically, this term is used in the context of user-defined functions, and it is mean to contrast with fenced functions, which execute in a separate process.  Unfenced functions are more efficient than fenced functions, but they are also more dangerous, as they can crash the process in which they are running.
+
+Unfenced functions may often be implemented to provide an efficient set of functions that are common to a business unit or group.  They allow the users of Presto to efficiently customize Presto into a different flavor specific to their business needs.
+
+A key feature of unfenced functions is that they can be used to implement constant folding.  This is because the function can be evaluated at query planning time, and the result can be used as a constant in the query plan.  This can be a significant performance improvement, as it can reduce the amount of work that needs to be done at query execution time.  For example, if an unfenced function is evaluated over a partition column, the result can be used to prune partitions before the query is executed.
+
+Currently in Presto C++, there is no constant folding support for functions which are written in C++.  In practice, constant folding works for functions which have a corresponding Java implementation, but not for functions which are implemented in C++.  It works by essentially just using a forked set of functions written in Java to evaluate the constant expressions, and then during execution relying on the C++ functions to evaluate the non-constant expressions.  This has a few drawbacks:
+
+1) It requires a forked set of functions to be written in Java, which is a maintenance burden.
+2) An unfenced function written in C++ must have a corresponding Java implementation in order to work properly.
+
+RFC-0003 proposed a mechanism to both add a function registry to allow C++ implemented functions to be exposed to the coordinator without a corresponding Java implementation, and to add constant folding support via remote function evaluation.  This RFC has a similar goal, but represents an improvement over that approach.
+
+In this RFC, we propose to optimize row expressions as a whole, not just function calls.  This has a few advantages:
+
+1) It allows for constant folding to be applied to expressions which are not function calls, such as special form expressions.
+2) It can reduce the overhead of constant folding by allowing the evaluation engine to evaluate the expression as a whole, rather than evaluating each function call individually.
+
+### [Optional] Goals
+
+* Allow for a pluggable implementation of expression evaluation and use this in the Presto optimizer for constant folding
+* Allow connector authors to use this plugin to optimize their own expressions (for example, for filter pushdown)
+
+### [Optional] Non-goals
```

Comment:
Should we mention that making this as performant and current approach is a non-goal? Or should we have some goals on "max" execution latency introduced by this proposal?